### ISSUE TITLE
Gradle Daemon client handles symlinked JAVA_HOME incorrectly

### ISSUE THREAD

**Issue by cab404** [gradle/gradle#11200]
I tried to run Idea's Gradle plugin with JVM from Nix and was surprised to get an error.
After several hours I finally found stacktrace.

``` log
org.gradle.tooling.GradleConnectionException: Could not run phased build action using Gradle distribution 'https://services.gradle.org/distributions/gradle-4.10.2-bin.zip'.
	at org.gradle.tooling.internal.consumer.ExceptionTransformer.transform(ExceptionTransformer.java:55)
	at org.gradle.tooling.internal.consumer.ExceptionTransformer.transform(ExceptionTransformer.java:29)
	at org.gradle.tooling.internal.consumer.ResultHandlerAdapter.onFailure(ResultHandlerAdapter.java:41)
	at org.gradle.tooling.internal.consumer.async.DefaultAsyncConsumerActionExecutor$1$1.run(DefaultAsyncConsumerActionExecutor.java:57)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.gradle.launcher.daemon.client.DaemonConnectionException: The newly created daemon process has a different context than expected.
It won't be possible to reconnect to this daemon. Context mismatch:
Java home is different.
Wanted: DefaultDaemonContext[uid=null,javaHome=/nix/store/icmpxqlriwkq36zlxcwj3ckins4pvpb7-openjdk-8u222-ga,daemonRegistryDir=/home/cab/gradlereset/daemon,pid=15270,idleTimeout=null,daemonOpts=-XX:+HeapDumpOnOutOfMemoryError,-Xmx1024m,-Dfile.encoding=US-ASCII,-Duser.country=US,-Duser.language=en,-Duser.variant]
Actual: DefaultDaemonContext[uid=8d7e6c5c-21de-4e9c-bc19-de185a3e62d8,javaHome=/nix/store/icmpxqlriwkq36zlxcwj3ckins4pvpb7-openjdk-8u222-ga/lib/openjdk,daemonRegistryDir=/home/cab/gradlereset/daemon,pid=17671,idleTimeout=10800000,daemonOpts=-XX:+HeapDumpOnOutOfMemoryError,-Xmx1024m,-Dfile.encoding=US-ASCII,-Duser.country=US,-Duser.language=en,-Duser.variant]

	at org.gradle.launcher.daemon.client.DefaultDaemonConnector.connectToDaemonWithId(DefaultDaemonConnector.java:246)
	at org.gradle.launcher.daemon.client.DefaultDaemonConnector.doStartDaemon(DefaultDaemonConnector.java:217)
	at org.gradle.launcher.daemon.client.DefaultDaemonConnector.startDaemon(DefaultDaemonConnector.java:206)
	at org.gradle.launcher.daemon.client.DefaultDaemonConnector.connect(DefaultDaemonConnector.java:130)
	at org.gradle.launcher.daemon.client.DaemonClient.execute(DaemonClient.java:139)
	at org.gradle.launcher.daemon.client.DaemonClient.execute(DaemonClient.java:93)
	at org.gradle.tooling.internal.provider.DaemonBuildActionExecuter.execute(DaemonBuildActionExecuter.java:60)
	at org.gradle.tooling.internal.provider.DaemonBuildActionExecuter.execute(DaemonBuildActionExecuter.java:41)
	at org.gradle.tooling.internal.provider.LoggingBridgingBuildActionExecuter.execute(LoggingBridgingBuildActionExecuter.java:58)
	at org.gradle.tooling.internal.provider.LoggingBridgingBuildActionExecuter.execute(LoggingBridgingBuildActionExecuter.java:37)
	at org.gradle.tooling.internal.provider.ProviderConnection.run(ProviderConnection.java:180)
	at org.gradle.tooling.internal.provider.ProviderConnection.runPhasedAction(ProviderConnection.java:157)
	at org.gradle.tooling.internal.provider.DefaultConnection.run(DefaultConnection.java:250)
	at org.gradle.tooling.internal.consumer.connection.PhasedActionAwareConsumerConnection.run(PhasedActionAwareConsumerConnection.java:56)
	at org.gradle.tooling.internal.consumer.connection.ParameterValidatingConsumerConnection.run(ParameterValidatingConsumerConnection.java:60)
	at org.gradle.tooling.internal.consumer.DefaultPhasedBuildActionExecuter$1.run(DefaultPhasedBuildActionExecuter.java:76)
	at org.gradle.tooling.internal.consumer.DefaultPhasedBuildActionExecuter$1.run(DefaultPhasedBuildActionExecuter.java:70)
	at org.gradle.tooling.internal.consumer.connection.LazyConsumerActionExecutor.run(LazyConsumerActionExecutor.java:84)
	at org.gradle.tooling.internal.consumer.connection.CancellableConsumerActionExecutor.run(CancellableConsumerActionExecutor.java:45)
	at org.gradle.tooling.internal.consumer.connection.ProgressLoggingConsumerActionExecutor.run(ProgressLoggingConsumerActionExecutor.java:58)
	at org.gradle.tooling.internal.consumer.connection.RethrowingErrorsConsumerActionExecutor.run(RethrowingErrorsConsumerActionExecutor.java:38)
	at org.gradle.tooling.internal.consumer.async.DefaultAsyncConsumerActionExecutor$1$1.run(DefaultAsyncConsumerActionExecutor.java:55)
	... 6 more
2019-11-01 19:53:04,580 [2190927]   INFO - System.util.ExternalSystemUtil - External project [/home/cab/data/work-lambda/bigsister-ktor] resolution task executed in 662 ms.
```

### Expected Behavior
JAVA_HOME check should first run `readlink` on required directories ('/bin', '/include', etc) on both paths, and check whether resolved paths match. Or JAVA_HOME check should be replaced with simpler, but probably less effective equality check of `readlink "$JAVA_HOME/bin/java"`.

### Current Behavior
JAVA_HOME check is dumb and checks whether two strings are equal.

### Context
Мismatching directories are effectively the same JAVA_HOME — bin, jre and include of one is symlinked to another!
``` log
[nix-shell:~]$ l /nix/store/95kq9w15v4amzgqyflfjh5jls0zxm91f-openjdk-8u222-ga
total 12K
dr-xr-xr-x 1 root root     64 Jan  1  1970 .
drwxrwxr-t 1 root nixbld 2.2M Nov  1 19:16 ..
lrwxrwxrwx 2 root root     76 Jan  1  1970 bin -> /nix/store/95kq9w15v4amzgqyflfjh5jls0zxm91f-openjdk-8u222-ga/lib/openjdk/bin
lrwxrwxrwx 2 root root     80 Jan  1  1970 include -> /nix/store/95kq9w15v4amzgqyflfjh5jls0zxm91f-openjdk-8u222-ga/lib/openjdk/include
lrwxrwxrwx 2 root root     80 Jan  1  1970 jre -> /nix/store/7fgnr4fw1zy3gvg6l09qhapxhrmqm06z-openjdk-8u222-ga-jre/lib/openjdk/jre
dr-xr-xr-x 1 root root     14 Jan  1  1970 lib
dr-xr-xr-x 1 root root     66 Jan  1  1970 nix-support
dr-xr-xr-x 1 root root      6 Jan  1  1970 share
```

Using actual directory resolved the issue, but not before eating 4 hours+ of ripping my hair off.

### Steps to Reproduce 
1. Install [IDEA](https://www.jetbrains.com/idea/download/#section=linux)
2. Install [Nix](https://nixos.org/nix/)
3. `nix-shell -p openjdk8`
4. `which java`, copy the `/nix/store/SOMEHASH-openjdk-8u222-ga` path
5. Open IDEA in created that shell 
6. Create template Gradle project
7. Open 'File | Project structure', select `/nix/store/SOMEHASH-openjdk-8u222-ga` as project interpreter
8. Open 'File | Settings | Build, Execution, Deployment | Build Tools | Gradle',  set 'Project interpreter' as Gradle JVM.
9. Try syncing project
10. Get Gradle Client Daemon error at `~/.Idea*/system/log/idea.log`

### Your Environment
I use NixOS 19.09, but this bug is reproducible on Nix + Ubuntu. Should probably be reproducible on any system supporting symlinks.

**Comment by big-guy** [comment]
The issue here is that the JAVA_HOMEs are different.  They just happen to have _some_ of the same content.

In your example, the requested JAVA_HOME was `/nix/store/95kq9w15v4amzgqyflfjh5jls0zxm91f-openjdk-8u222-ga` and the actual one used was `/nix/store/7fgnr4fw1zy3gvg6l09qhapxhrmqm06z-openjdk-8u222-ga-jre/lib/openjdk`.  When comparing these paths, we do resolve any symlinks, but the first path is not a symlink to the second one.

A Java home has more than just bin, include and jre.  The lib directory has other important things in it and that directory isn't symlinked, so I'm not sure how we would know if two different paths were equivalent without deeper inspection of the contents.   There are also optional directories and files that may or may not exist.

As a workaround, is it possible to use `/nix/store/95kq9w15v4amzgqyflfjh5jls0zxm91f-openjdk-8u222-ga/lib/openjdk` as your JAVA_HOME?  Does `/nix/store/95kq9w15v4amzgqyflfjh5jls0zxm91f-openjdk-8u222-ga/lib/openjdk/bin` symlink to `/nix/store/7fgnr4fw1zy3gvg6l09qhapxhrmqm06z-openjdk-8u222-ga-jre/lib/openjdk/bin`?  If so, I don't think this will work either.

**Comment by cab404** [comment]
> As a workaround, is it possible to use `/nix/store/95kq9w15v4amzgqyflfjh5jls0zxm91f-openjdk-8u222-ga/lib/openjdk` as your JAVA_HOME?

That works, and at time of writing this issue I used that as a workaround. (sorry for not submitting it into this issue)


> The issue here is that the JAVA_HOMEs are different. They just happen to have _some_ of the same content.

Then issue might be in JAVA_HOME resolution in either IDEA or Gradle daemon launcher — I did not gave different paths for daemon and client, after all)
Not sure yet, which one is to blame. Does Gradle daemon do some funky stuff to get it's JAVA_HOME?

**Comment by big-guy** [comment]
The one we get back from the daemon is the one that Java reports as `System.getProperty("java.home")`.

**Comment by cab404** [comment]
> The one we get back from the daemon is the one that Java reports as `System.getProperty("java.home")`.

And DaemonContext it compared to [uses](https://github.com/gradle/gradle/blob/4e29e9035463eaa44ec4d657ed4cb7f7e07b5d33/subprojects/launcher/src/main/java/org/gradle/launcher/daemon/context/DaemonContext.java#L47) JAVA_HOME. 

I did a simple test, and usually JAVA_HOME != java.home

``` java
public class Test {
public static void main(String... args) {
	System.out.println("java.home:" + System.getProperty("java.home"));
	System.out.println("JAVA_HOME:" + System.getenv("JAVA_HOME"));
}}
``` 

1. `nix-shell -p openjdk8`
2. `javac Test.java`
3. `java Test`

```
[nix-shell:~]$ /nix/store/95kq9w15v4amzgqyflfjh5jls0zxm91f-openjdk-8u222-ga/bin/java Test
java.home:/nix/store/95kq9w15v4amzgqyflfjh5jls0zxm91f-openjdk-8u222-ga/lib/openjdk/jre
JAVA_HOME:/nix/store/95kq9w15v4amzgqyflfjh5jls0zxm91f-openjdk-8u222-ga/lib/openjdk
```

As you can see, `env()` call does a readlink or something like that, so it shouldn't be particularly reliable to compare that to JAVA_HOME. Maybe use env on both ends?

**Comment by soloturn** [comment]
@cab404 i think @big-guy is right: java knows where it is installed, requiring, setting, and using JAVA_HOME should not be done, see also https://stackoverflow.com/questions/1117398/java-home-directory-in-linux , also for setting the variable on command line fjor linux/windows. this might be a nix linux bug. or, your profile is wrong.

if gradle or gradlew should report that JAVA_HOME is different to where java thinks it is installed is a different question, one might argue that it should prefer java retrieved out of the path and remove other complexities out of its code.

e.g. on arch linux  JAVA_HOME is not set:
```
$ java Test
java.home:/usr/lib/jvm/java-8-openjdk/jre
JAVA_HOME:null
```
if some program really insists that JAVA_HOME is set, one could grep and sed on the command line:
```
$ java -XshowSettings:properties -version
$ java -XshowSettings:properties -version 2>&1 | grep "java.home"
    java.home = /usr/lib/jvm/java-8-openjdk/jre
```

**Comment by cab404** [comment]
> if gradle or gradlew should report that JAVA_HOME is different to where java thinks it is installed is a different question, one might argue that it should prefer java retrieved out of the path and remove other complexities out of its code.

The problem is that I used the same java executable in both runs.That might mean that _gradle_ is doing something silly to get it's java_home path.

**Comment by soloturn** [comment]
@cab404 is this still a problem? nix is here like arch, and on arch intellij, gradle works. no JAVA_HOME set. see the openjdk and idea packages, as well how they change the default:
* https://archlinux.org/packages/community/x86_64/intellij-idea-community-edition/
* https://archlinux.org/packages/extra/x86_64/jdk-openjdk/
* https://wiki.archlinux.org/title/java and "sudo archlinux-java set java-11-openjdk" - which is a little simplistic as, done as root, it changes the default for all users

**Comment by ljacomet** [comment]
Could you see if the fix for #12840 which has been fixed in 6.9+ fixes this issue?

**Comment by ljacomet** [comment]
Closing as this should have been fixed by #12840

**Comment by emartynov** [comment]
Hi people, I think I still have similar issue:

I set JAVA_HOME to ~/.sdkman/candidates/java/current, which is symbolic link. However, AS and IJ are resolving it to the current set SDK. I saw that when I run gradle from command line it is starting new daemon. And it looks that literal difference between symbolic link and resolved folder is the reason. I tried to use gradle doctor plugin and got error
```
| Gradle is not using JAVA_HOME.                                                                       |
| JAVA_HOME is ~/.sdkman/candidates/java/current                                            |
| Gradle is using ~/.sdkman/candidates/java/15.0.5-zulu/zulu-15.jdk/Contents/Home  
```

So I did gradle --stop. I run gradle build in AS. I opened terminal and just run gradle tasks in it. I saw a new daemon start. I also see that I have two daemons now.
